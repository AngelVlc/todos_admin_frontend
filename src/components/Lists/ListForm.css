.dnd-container {
  margin: 8px;
}

.dnd-list {
  padding: 8px;
}

.list-button {
  padding: 8px;
}

.dnd-item {
  border: 1px solid lightgrey;
  border-radius: 2px;
  padding: 8px;
  margin-bottom: 8px;
}