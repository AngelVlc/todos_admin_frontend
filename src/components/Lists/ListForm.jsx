import React, { useEffect, useState } from 'react';
import { useHistory, Link } from 'react-router-dom';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import axios from 'axios';
import * as Yup from 'yup';

export const ListForm = (props) => {
  let history = useHistory();

  const [pageState, setPageState] = useState({
    name: '',
    items: []
  });

  useEffect(() => {
    if (props.name) {
        setPageState({
            name: props.name,
            items: props.items
        })
    }
  }, [props]);

  const onSubmit = async ({ name }) => {
    const body = {
        name
    }
    let res;
    if (props.isNew) {
        res = await axios.post(props.submitUrl, body)
    } else {
        res = await axios.put(props.submitUrl, body)
    }
    history.push(`/lists/${res.data.id}/edit`);
  };

  return (
    <Formik
        enableReinitialize={true}
        initialValues={pageState}
        validationSchema={
            Yup.object({
                name: Yup.string()
                    .required('Required')
            })
        }
        onSubmit={onSubmit}>
        <Form>
            <div className="field">
                <label className="label" htmlFor="name">Name</label>
                <div className="control">
                    <Field name="name" type="text" data-testid="name" />
                </div>
                <p className="help is-danger" data-testid="nameErrors">
                    <ErrorMessage name="name" />
                </p>
            </div>
            {!props.isNew &&
                <table className="table">
                    <thead>
                        <tr>
                            <td>Title</td>
                            <td>

                                <button className="button is-small" type="button" data-testid="addNew" onClick={() => history.push(`/lists/${props.listId}/items/new`)}>
                                    <span className="icon is-small">
                                        <i className="fas fa-plus"></i>
                                    </span>
                                </button>
                            </td>
                        </tr>
                    </thead>
                    <tbody>
                        {props.items.length > 0 && props.items.map((item) => (
                            <tr key={item.id}>
                                <td>
                                    <Link className="has-text-black" data-testid={`editListItem${item.id}`} to={`/lists/${props.listId}/items/${item.id}/edit`}>{item.title}</Link>
                                </td>
                                <td>
                                    <center>
                                        <Link className="has-text-black" data-testid={`deleteListItem${item.id}`} to={`/lists/${props.listId}/items/${item.id}/delete`}>
                                            <span className="icon is-small">
                                                <i className="fas fa-trash-alt fa-xs"></i>
                                            </span>
                                        </Link>
                                    </center>
                                </td>
                            </tr>
                        ))
                        }
                    </tbody>
                </table>
            }

            <div className="field is-grouped">
                <p className="control">
                    <button className="button" data-testid="submit" type="submit">{props.submintBtnText}</button>
                </p>
                <p className="control">
                    <button className="button" data-testid="cancel" type="button" onClick={() => history.push('/lists')}>CANCEL</button>
                </p>
                {!props.isNew &&
                    <p className="control">
                        <button className="button is-danger" data-testid="delete" type="button" onClick={() => history.push(`/lists/${props.listId}/delete`)}>DELETE</button>
                    </p>
                }
            </div>
        </Form>
    </Formik>
  )
}
