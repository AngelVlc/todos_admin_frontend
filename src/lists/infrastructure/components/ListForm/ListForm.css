.dnd-item {
  border: 1px solid lightgrey;
  border-radius: 2px;
}

.rounded-button {
  background-color: rgba(10, 10, 10, 0.2);
  border: none;
  border-radius: 9999px;
  color: white;
  height: 20px;
  width: 20px;
  vertical-align: top;
}
