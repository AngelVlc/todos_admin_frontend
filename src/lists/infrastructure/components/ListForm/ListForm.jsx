import React, { useState, useContext, useRef } from "react";
import { useHistory } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";
import * as Yup from "yup";
import "./ListForm.css";
import { List, ListItem } from "../../../domain";
import {
  CreateListUseCase,
  UpdateListUseCase,
} from "../../../application/lists";
import { Modal } from "../../../../shared/infrastructure/components/Modal";
import { ListItemForm } from "../ListItemForm";

export const ListForm = (props) => {
  let history = useHistory();
  const { useCaseFactory } = useContext(AppContext);
  const itemFormModalRef = useRef();
  const itemFormRef = useRef();

  const [pageState, setPageState] = useState(props.list);

  const onSubmit = async (list) => {
    let useCase;

    if (props.list?.id === -1) {
      useCase = useCaseFactory.get(CreateListUseCase);
    } else {
      useCase = useCaseFactory.get(UpdateListUseCase);
    }

    const result = await useCase.execute(list);

    if (result && props.list?.id === -1) {
      history.push(`/lists/${result.id}`);
    }
  };

  const onDragEnd = (dropResult) => {
    const { destination, source } = dropResult;

    if (!destination) {
      return;
    }

    if (
      destination.droppableId === source.droppableId &&
      destination.index === source.index
    ) {
      return;
    }

    const newItems = Array.from(pageState.items);
    newItems.splice(source.index, 1);
    newItems.splice(destination.index, 0, pageState.items[source.index]);

    const newList = new List({ ...pageState, items: newItems });

    setPageState(newList);
  };

  const onDeleteListItem = (item) => {
    const index = pageState.items.indexOf(item);
    const newItems = Array.from(pageState.items);
    newItems[index].state = "deleted";

    const newList = new List({ ...pageState, items: newItems });

    setPageState(newList);
  };

  const onAddNewItem = () => {
    showItemModal(ListItem.createEmpty(props.list?.id));
  };

  const onEditItem = (item) => {
    showItemModal(item);
  };

  const showItemModal = (item) => {
    itemFormRef.current.setValues(item);
    itemFormModalRef.current.showModal();
  };

  const onSubmitItemForm = (listItem) => {
    const newList = new List({ ...pageState, items: pageState.items });

    if (listItem.id === -1) {
      newList.addNewItem(listItem);
    } else {
      for (const item of newList.items) {
        if (item.id === listItem.id) {
          item.title = listItem.title;
          item.description = listItem.description;
          item.state = "modified";
          item.listId = listItem.listId;

          break;
        }
      }
    }

    setPageState(newList);

    itemFormModalRef.current.closeModal();
  };

  const onItemModalClose = () => {
    itemFormRef.current.submitForm();
  };

  return (
    <>
      <Modal ref={itemFormModalRef} closeHandler={onItemModalClose}>
        <ListItemForm ref={itemFormRef} onSubmit={onSubmitItemForm} />
      </Modal>
      <Formik
        enableReinitialize={true}
        initialValues={pageState}
        validationSchema={Yup.object({
          name: Yup.string().required("Required"),
        })}
        onSubmit={onSubmit}
      >
        <Form>
          <div className="field">
            <label className="label" htmlFor="name">
              Name
            </label>
            <div className="control">
              <Field
                name="name"
                as="input"
                className="input"
                data-testid="name"
                autoFocus
              />
            </div>
            <p className="help is-danger" data-testid="nameErrors">
              <ErrorMessage name="name" />
            </p>
          </div>
          <div>
            <div className="is-flex">
              <div className="is-flex-grow-4">
                <span className="label">List Items</span>
              </div>
            </div>
            <DragDropContext onDragEnd={onDragEnd}>
              <div className="dnd-container">
                <div className="list-button mt-2 mb-2">
                  <button
                    className="button is-small"
                    type="button"
                    data-testid="addNew"
                    onClick={() => onAddNewItem()}
                  >
                    <span className="icon is-small">
                      <i className="fas fa-plus"></i>
                    </span>
                    <span>Add</span>
                  </button>
                </div>
                <Droppable droppableId={props.list.id.toString()}>
                  {(provided) => (
                    <div
                      className="dnd-list"
                      ref={provided.innerRef}
                      {...provided.droppableProps}
                    >
                      {pageState.items.length > 0 &&
                        pageState.items
                          .filter((item) => item.state !== "deleted")
                          .map((item, index) => (
                            <Draggable
                              key={item.id}
                              draggableId={item.id.toString()}
                              index={index}
                            >
                              {(draggableProvided) => (
                                <div
                                  key={item.id}
                                  className="is-flex dnd-item p-2 mb-2"
                                  {...draggableProvided.draggableProps}
                                  {...draggableProvided.dragHandleProps}
                                  ref={draggableProvided.innerRef}
                                  data-testid={`draggable${item.id}`}
                                >
                                  <div
                                    className="is-flex-grow-4"
                                    data-testid={`editListItem${item.id}`}
                                    onClick={() => onEditItem(item)}
                                  >
                                    <span className="has-text-black">
                                      {item.title}
                                    </span>
                                  </div>
                                  <div className="is-justify-content-flex-end">
                                    <center>
                                      <button
                                        type="button"
                                        className="has-text-black delete"
                                        data-testid={`deleteListItem${item.id}`}
                                        onClick={() => onDeleteListItem(item)}
                                      ></button>
                                    </center>
                                  </div>
                                </div>
                              )}
                            </Draggable>
                          ))}
                      {provided.placeholder}
                    </div>
                  )}
                </Droppable>
              </div>
            </DragDropContext>
          </div>
          <div className="field is-grouped">
            <div className="control">
              <button className="button" data-testid="submit" type="submit">
                {props.list?.id ? "SAVE" : "CREATE"}
              </button>
            </div>
            <>{props.preCancel}</>
            <div className="control ml-auto is-pulled-right">
              <button
                className="button"
                data-testid="cancel"
                type="button"
                onClick={() => history.push("/lists")}
              >
                CANCEL
              </button>
            </div>
          </div>
        </Form>
      </Formik>
    </>
  );
};
