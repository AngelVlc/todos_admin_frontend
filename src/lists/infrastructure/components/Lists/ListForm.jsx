import React, { useState, useContext } from "react";
import { useHistory, Link } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";
import * as Yup from "yup";
import "./ListForm.css";
import { List } from "../../../domain";
import {
  CreateListUseCase,
  UpdateListUseCase,
} from "../../../application/lists";

export const ListForm = (props) => {
  let history = useHistory();
  const { useCaseFactory } = useContext(AppContext);

  const [pageState, setPageState] = useState(props.list);

  const onSubmit = async (list) => {
    let useCase;
    if (props.list?.id === undefined) {
      useCase = useCaseFactory.get(CreateListUseCase);
    } else {
      useCase = useCaseFactory.get(UpdateListUseCase);
    }

    const result = await useCase.execute(list);
    if (result && props.list?.id === undefined) {
      history.push(`/lists/${result.id}`);
    }
  };

  const onDragEnd = (dropResult) => {
    const { destination, source } = dropResult;

    if (!destination) {
      return;
    }

    if (
      destination.droppableId === source.droppableId &&
      destination.index === source.index
    ) {
      return;
    }

    const newItems = Array.from(pageState.items);
    newItems.splice(source.index, 1);
    newItems.splice(destination.index, 0, pageState.items[source.index]);

    setPageState(new List({ ...pageState, items: newItems }));
  };

  return (
    <Formik
      enableReinitialize={true}
      initialValues={pageState}
      validationSchema={Yup.object({
        name: Yup.string().required("Required"),
      })}
      onSubmit={onSubmit}
    >
      <Form>
        <div className="field">
          <label className="label" htmlFor="name">
            Name
          </label>
          <div className="control">
            <Field
              name="name"
              as="input"
              className="input"
              data-testid="name"
              autoFocus
            />
          </div>
          <p className="help is-danger" data-testid="nameErrors">
            <ErrorMessage name="name" />
          </p>
        </div>
        <div className="field">
          <label className="label" htmlFor="isQuickList">
            Is Quick List
          </label>
          <div className="control">
            <Field
              name="isQuickList"
              type="checkbox"
              data-testid="isQuickList"
            />
          </div>
        </div>
        {props.list?.id && (
          <div>
            <div className="is-flex">
              <div className="is-flex-grow-4">
                <span className="label">List Items</span>
              </div>
            </div>
            <DragDropContext onDragEnd={onDragEnd}>
              <div className="dnd-container">
                <div className="list-button mt-2 mb-2">
                  <button
                    className="button is-small"
                    type="button"
                    data-testid="addNew"
                    onClick={() =>
                      history.push(`/lists/${props.list.id}/items/new`)
                    }
                  >
                    <span className="icon is-small">
                      <i className="fas fa-plus"></i>
                    </span>
                    <span>Add</span>
                  </button>
                </div>
                <Droppable droppableId={props.list.id.toString()}>
                  {(provided) => (
                    <div
                      className="dnd-list"
                      ref={provided.innerRef}
                      {...provided.droppableProps}
                    >
                      {pageState.items.length > 0 &&
                        pageState.items.map((item, index) => (
                          <Draggable
                            key={item.id}
                            draggableId={item.id.toString()}
                            index={index}
                          >
                            {(draggableProvided) => (
                              <div
                                key={item.id}
                                className="is-flex dnd-item p-2 mb-2"
                                {...draggableProvided.draggableProps}
                                {...draggableProvided.dragHandleProps}
                                ref={draggableProvided.innerRef}
                                data-testid={`draggable${item.id}`}
                              >
                                <Link
                                  className="is-flex-grow-4"
                                  data-testid={`editListItem${item.id}`}
                                  to={`/lists/${props.list.id}/items/${item.id}/edit`}
                                >
                                  <div>
                                    <span className="has-text-black">
                                      {item.title}
                                    </span>
                                  </div>
                                </Link>
                                <div className="is-justify-content-flex-end">
                                  <center>
                                    <Link
                                      className="has-text-black delete"
                                      data-testid={`deleteListItem${item.id}`}
                                      to={`/lists/${props.list.id}/items/${item.id}/delete`}
                                    ></Link>
                                  </center>
                                </div>
                              </div>
                            )}
                          </Draggable>
                        ))}
                      {provided.placeholder}
                    </div>
                  )}
                </Droppable>
              </div>
            </DragDropContext>
          </div>
        )}

        <div className="field is-grouped">
          <div className="control">
            <button className="button" data-testid="submit" type="submit">
              {props.list?.id ? "SAVE" : "CREATE"}
            </button>
          </div>
          {props.list?.id && (
            <div className="control">
              <button
                className="button"
                data-testid="read"
                type="button"
                onClick={() => history.push(`/lists/${props.list.id}/read`)}
              >
                READ
              </button>
            </div>
          )}
          <div className="control">
            <button
              className="button"
              data-testid="cancel"
              type="button"
              onClick={() => history.push("/lists")}
            >
              CANCEL
            </button>
          </div>
          {props.list?.id && (
            <div className="control ml-auto is-pulled-right">
              <button
                className="button is-danger"
                data-testid="delete"
                type="button"
                onClick={() => history.push(`/lists/${props.list.id}/delete`)}
              >
                DELETE
              </button>
            </div>
          )}
        </div>
      </Form>
    </Formik>
  );
};
