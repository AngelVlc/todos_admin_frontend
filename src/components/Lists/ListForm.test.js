import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { ListForm } from './ListForm'
import { AppContext } from '../../contexts/AppContext'
import axios from 'axios';
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils';

jest.mock('axios');
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush
    })
}));

const renderWithContextAndRouterForExistingList = () => {
    const items = [
                {
                    id: 5,
                    title: 'item title',
                    description: 'item description'
                }
            ]

    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/lists/2/items/5/edit`]}>
                    <Route path="/lists/:listId/items/:itemId/edit">
                      <ListForm listId={2} name='list name' items={items} isNew={false} submintBtnText='SAVE' submitUrl={'lists/2'}/>
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

const renderWithContextAndRouterForNewList = () => {
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/users/new`]}>
                    <Route path="/users/new">
                      <ListForm name='' items={[]} isNew={true} submintBtnText='CREATE' submitUrl={'lists'}/>
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should match the snapshot for an existing list', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouterForExistingList();
        fragment = asFragment;
    });
    expect(fragment()).toMatchSnapshot();
});

it('should match the snapshot for a new list', async () => {
    const { asFragment } = renderWithContextAndRouterForNewList();
    expect(asFragment()).toMatchSnapshot();
});

it('should allow delete an existing list', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingList();
    });

    await wait(() => {
        fireEvent.click(container.getByTestId('delete'));
    })

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/lists/2/delete');
    mockHistoryPush.mockClear();
});

it('should allow cancel', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewList();

    await wait(() => {
        fireEvent.click(getByTestId('cancel'));
    })

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/lists');
    mockHistoryPush.mockClear();
});

it('should require list name', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewList();

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(getByTestId('nameErrors')).toHaveTextContent('Required');
})

it('should update an existing list', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingList();
    });

    await changeInputValue(container.getByTestId, 'name', 'updated name');

    axios.put.mockResolvedValue({ data: {id: 2} });

    await wait(() => {
        fireEvent.click(container.getByTestId('submit'));
    })

    expect(axios.put.mock.calls.length).toBe(1);
    expect(axios.put.mock.calls[0][0]).toBe('lists/2');
    expect(axios.put.mock.calls[0][1]).toStrictEqual({ name: 'updated name' });

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe(`/lists/2/edit`);
    mockHistoryPush.mockClear();
});

it('should create a new list', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewList();

    await changeInputValue(getByTestId, 'name', 'new list');

    axios.post.mockResolvedValue({ data: {id: 55} });

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(axios.post.mock.calls.length).toBe(1);
    expect(axios.post.mock.calls[0][0]).toBe('lists');
    expect(axios.post.mock.calls[0][1]).toStrictEqual({ name: 'new list' });

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/lists/55/edit');
    mockHistoryPush.mockClear();
});

it('should add a new item', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingList();
    });

    await wait(() => {
        fireEvent.click(container.getByTestId('addNew'));
    })

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/lists/2/items/new');
    mockHistoryPush.mockClear();
});

it('should update an existing item', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingList();
    });

    expect(container.getByTestId('editListItem5').href).toBe('http://localhost/lists/2/items/5/edit');
});

it('should allow delete an existing item item', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingList();
    });

    expect(container.getByTestId('deleteListItem5').href).toBe('http://localhost/lists/2/items/5/delete');
});

const changeInputValue = async (getByTestId, name, value) => {
    await wait(() => {
        fireEvent.change(getByTestId(name), {
            target: {
                value: value
            }
        })
    })
}
