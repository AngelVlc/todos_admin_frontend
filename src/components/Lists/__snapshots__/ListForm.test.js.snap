// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for a new list 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="name"
      >
        Name
      </label>
      <div
        class="control"
      >
        <input
          data-testid="name"
          name="name"
          type="text"
          value=""
        />
      </div>
      <p
        class="help is-danger"
        data-testid="nameErrors"
      />
    </div>
    <div
      class="field is-grouped"
    >
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          CREATE
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </p>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing list 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="name"
      >
        Name
      </label>
      <div
        class="control"
      >
        <input
          data-testid="name"
          name="name"
          type="text"
          value="list name"
        />
      </div>
      <p
        class="help is-danger"
        data-testid="nameErrors"
      />
    </div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <td>
            Title
          </td>
          <td>
            <button
              class="button is-small"
              data-testid="addNew"
              type="button"
            >
              <span
                class="icon is-small"
              >
                <i
                  class="fas fa-plus"
                />
              </span>
            </button>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              class="has-text-black"
              data-testid="editListItem5"
              href="/lists/2/items/5/edit"
            >
              item title
            </a>
          </td>
          <td>
            <center>
              <a
                class="has-text-black"
                data-testid="deleteListItem5"
                href="/lists/2/items/5/delete"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="fas fa-trash-alt fa-xs"
                  />
                </span>
              </a>
            </center>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="field is-grouped"
    >
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          SAVE
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button is-danger"
          data-testid="delete"
          type="button"
        >
          DELETE
        </button>
      </p>
    </div>
  </form>
</DocumentFragment>
`;
